code/datums/components/ranged_mob_full_auto.dm

```diff
@@ -0,0 +1,203 @@
+	RegisterSignal(parent, COMSIG_MOB_LOGIN, PROC_REF(on_gained_client))
+	RegisterSignal(parent, COMSIG_MOB_LOGOUT, PROC_REF(on_lost_client))
+
+/datum/component/ranged_mob_full_auto/UnregisterFromParent()
+	UnregisterSignal(parent, list(COMSIG_MOB_LOGIN, COMSIG_MOB_LOGOUT))
+
+/datum/component/ranged_mob_full_auto/process(seconds_per_tick)
+	if (!try_shooting())
+		return PROCESS_KILL
+
+/// Try and take a shot, returns false if we are unable to do so and should stop trying
+/datum/component/ranged_mob_full_auto/proc/try_shooting()
+	if (!is_firing)
+		return FALSE
+	if (!COOLDOWN_FINISHED(src, next_shot_cooldown))
+		return TRUE // Don't fire but also keep processing
+
+	var/mob/living/living_parent = parent
+
+	if (isnull(target) || get_turf(target) != target_loc) // Target moved or got destroyed since we last aimed.
+		set_target(target_loc)
+		target = target_loc // So we keep firing on the emptied tile until we move our mouse and find a new target.
+	if (get_dist(living_parent, target) <= 0)
+		set_target(get_step(living_parent, living_parent.dir)) // Shoot in the direction faced if the mouse is on the same tile as we are.
+		target_loc = target
+	else if (!in_view_range(living_parent, target))
+		stop_firing()
+		return FALSE // Can't see shit
+
+	living_parent.face_atom(target)
+	COOLDOWN_START(src, next_shot_cooldown, autofire_shot_delay)
+	living_parent.RangedAttack(target)
+	return TRUE
+
+/// Setter for reference handling
+/datum/component/ranged_mob_full_auto/proc/set_target(atom/new_target)
+	if (!isnull(target))
+		UnregisterSignal(target, COMSIG_QDELETING)
+	target = new_target
+	if (!isnull(target))
+		RegisterSignal(target, COMSIG_QDELETING, PROC_REF(on_target_deleted))
+
+/// Don't hang references
+/datum/component/ranged_mob_full_auto/proc/on_target_deleted()
+	SIGNAL_HANDLER
+	set_target(null)
+
+/// When we gain a client, start tracking clicks
+/datum/component/ranged_mob_full_auto/proc/on_gained_client(mob/living/source)
+	SIGNAL_HANDLER
+	clicker = source.client
+	RegisterSignal(clicker, COMSIG_CLIENT_MOUSEDOWN, PROC_REF(on_mouse_down))
+
+/// When we lose our client, stop functioning
+/datum/component/ranged_mob_full_auto/proc/on_lost_client(mob/living/source)
+	SIGNAL_HANDLER
+	if (!isnull(clicker))
+		UnregisterSignal(clicker, list(COMSIG_CLIENT_MOUSEDOWN, COMSIG_CLIENT_MOUSEDRAG, COMSIG_CLIENT_MOUSEUP))
+	stop_firing()
+	clicker = null
+
+/// On mouse down start shooting!
+/datum/component/ranged_mob_full_auto/proc/on_mouse_down(client/source, atom/target, turf/location, control, params)
+	SIGNAL_HANDLER
+	if (awaiting_status != AUTOFIRE_MOUSEDOWN)
+		return // Avoid a double mousedown with no mouseup
+	var/list/modifiers = params2list(params)
+
+	if (LAZYACCESS(modifiers, SHIFT_CLICK))
+		return
+	if (LAZYACCESS(modifiers, CTRL_CLICK))
+		return
+	if (LAZYACCESS(modifiers, MIDDLE_CLICK))
+		return
+	if (LAZYACCESS(modifiers, RIGHT_CLICK))
+		return
+	if (LAZYACCESS(modifiers, ALT_CLICK))
+		return
```

Comment:
we really need to have a `LAZYACCESS()` proc-like helper that can check for a match in a list of stuff this is insane